requirements: Upgrade boto to boto3.

whoodes opened the pull request:
Working off the chain of #10986, and #10370.

Fixes: #3490

whoodes commented on zerver/lib/upload.py:
```diff
@@ -322,33 +311,51 @@ def get_file_info(request: HttpRequest, user_file: File) -> Tuple[str, int, Opti
+            s3_object = get_s3_object(bucket_name, path_id)
+            # Call `HEAD object` to see if the object exists
+            s3_object.load()
+        except Exception:
```
So sometimes we see a `ClientError`, other times we might see a `NoSuchBucket` error.  However actually  specifying the exception is super weird as explained in the GH issue mentioned.  Not sure of the best way to deal with this.

whoodes commented on zerver/lib/upload.py:
```diff
@@ -322,33 +311,51 @@ def get_file_info(request: HttpRequest, user_file: File) -> Tuple[str, int, Opti
     def __init__(self) -> None:
-        self.connection = S3Connection(settings.S3_KEY, settings.S3_SECRET_KEY)
+        self.session = boto3.Session(settings.S3_KEY, settings.S3_SECRET_KEY,
+                                     region_name='s3.amazonaws.com')
```
I'm not sure if setting `region_name` is the right call here.  But it appears boto3 doesn't have a `DefaultHost` equivalent in `Session`.  As it seems it now requires you to specify the default.  Which we can do here, and then an `aws_config` could override that if needed I'm guessing? Not sure.

whoodes commented on zerver/tests/test_import_export.py:
```diff
@@ -92,10 +93,6 @@
     get_huddle_hash,
 )
 
-from zerver.lib.test_helpers import (
```
This was a duplicate so I just went ahead and got rid of this while I was at it.

whoodes commented on zerver/lib/upload.py:
```diff
@@ -600,24 +599,16 @@ def get_emoji_url(self, emoji_file_name: str, realm_id: int) -> str:
-            key=key.key)
+        key = os.path.join("exports", generate_random_token(32), os.path.basename(tarball_path))
+        s3_object = get_s3_object(settings.S3_AVATAR_BUCKET, key)
+        s3_object.upload_file(tarball_path)
```
It looks like that silent failure was [possibly resolved already](https://github.com/boto/boto3/issues/1067#issuecomment-594199212).

whoodes commented on zerver/tests/test_transfer.py:
```diff
@@ -41,18 +42,18 @@ def test_transfer_avatars_to_s3(self) -> None:
-        self.assertEqual(image_key.get_contents_as_string(), open(avatar_disk_path(user), "rb").read())
-        self.assertEqual(original_image_key.get_contents_as_string(), open(avatar_disk_path(user, original=True), "rb").read())
-        self.assertEqual(medium_image_key.get_contents_as_string(), open(avatar_disk_path(user, medium=True), "rb").read())
+        self.assertEqual(sum(1 for _ in bucket.objects.all()), 3)
```
`bucket.objects.all()` doesn't have `__len__` defined.

whoodes commented on zerver/tests/test_upload.py:
```diff
@@ -1574,9 +1576,12 @@ def test_file_upload_authed(self) -> None:
-        redirect_url = response['Location']
+        self.assertEqual(response.status_code, 302)
 
-        self.assertEqual(b"zulip!", urllib.request.urlopen(redirect_url).read().strip())
```
Here's the `403` error I mentioned in chat. I couldn't find a better way to get keys from reading the docs, other than iterating through them.

whoodes commented on zerver/tests/test_upload.py:
```diff
@@ -1598,24 +1603,26 @@ def test_upload_avatar_image(self) -> None:
+        # Boto3 requires us to iterate through the keys
+        for key in bucket.objects.all():
+            if key is medium_image_key:
+                key.delete()  # nocoverage
```
I wasn't sure how to get coverage in this case?

whoodes commented on zerver/tests/test_upload.py:
```diff
@@ -1673,16 +1679,21 @@ def test_delete_avatar_image(self) -> None:
+        with self.assertRaises(botocore.exceptions.ClientError):
+            bucket.Object(avatar_original_image_path_id).load()
+        with self.assertRaises(botocore.exceptions.ClientError):
+            bucket.Object(avatar_medium_path_id).load()
```
This probably isn't the right way, but I also lost coverage having this in a single `with` statement.  At the moment, I can't think of a better way to check that the keys are no longer there.

andersk replied:
A single `with` statement wouldn’t be right anyway. That would only test that one of the three lines raises an exception.

whoodes commented on zerver/lib/upload.py:
```diff
@@ -10,9 +10,8 @@
-from boto.s3.key import Key
-from boto.s3.connection import S3Connection
+import boto3
+
```
Woops. I'll get rid of the extra white space here.

whoodes commented on zerver/tests/test_upload.py:
```diff
@@ -1789,7 +1799,7 @@ def test_tarball_upload_and_deletion(self) -> None:
         if result is not None:
             hex_value = result.group(1)
-        expected_url = "https://{bucket}.s3.amazonaws.com:443/exports/{hex_value}/{path}".format(
+        expected_url = "https://{bucket}.s3.amazonaws.com/exports/{hex_value}/{path}".format(
```
It's not clear to my why this port number is removed now.  But testing this on the dev server, the download works and contains the contents as expected.

andersk commented on requirements/common.in:
```diff
@@ -30,7 +30,7 @@ SQLAlchemy
 
 # Needed for S3 file uploads
-boto
+boto3==1.9.240
```
Just `boto3`; don’t pin a specific version here (especially one that causes downgrades).

andersk commented on requirements/dev.txt:
```diff
@@ -66,16 +66,16 @@ beautifulsoup4==4.8.2 \
-    # via aws-sam-translator, moto
+boto3==1.9.240 \
+    --hash=sha256:0895805810df8cc49898fbeb61479c20c63a59fb6d3adf19f78fde90efec10a2 \
+    --hash=sha256:2f174f1e01d898ed940677a66404b7d30cbe92665b1fa2a5086eb64382ce2c41
```
You should `git checkout upstream requirements/*.txt; tools/update-locked-requirements` to get rid of these downgrades.

whoodes replied:
Cool thanks I was just looking at that and wondering.

whoodes commented on zerver/lib/test_helpers.py:
```diff
@@ -465,9 +464,10 @@ def new_method(*args: Any, **kwargs: Any) -> Any:
-def create_s3_buckets(*bucket_names: Tuple[str]) -> List[Bucket]:
-    conn = S3Connection(settings.S3_KEY, settings.S3_SECRET_KEY)
-    buckets = [conn.create_bucket(name) for name in bucket_names]
+def create_s3_buckets(*bucket_names: Tuple[str]) -> List[Any]:
```
I almost forgot to mention; I was also having trouble importing ‘Bucket’ via ‘boto3.Session.resource’ to use as a the return List type here for some reason. I’d get errors stating “...has no module Bucket”.

whoodes replied:
(Which was confusing  because it’s the approach I used to annotate the buckets in within the tests for mypy.)

whoodes replied:
https://github.com/python/typeshed/issues/2706 is relevant here.

whoodes replied:
(Just changed this to be a `ServiceResource` type, along with the buckets in `test_upload`)

timabbott commented:
So, this PR has a lot of complexity in the upgrade commit that doesn't need to be there.  E.g. in `zerver/lib/export.py`, we rename the `key` local variable to `sro` which changes tons of lines.  First, what's the reasoning behind that change in naming -- is that an upstream change, and is `sro` the right name, or something like `s3_resource`?  And in any case, I think we'd be a lot better off doing that rename in a separate commit either before or after the big version upgrade.

whoodes commented:
> So, this PR has a lot of complexity in the upgrade commit that doesn't need to be there. E.g. in `zerver/lib/export.py`, we rename the `key` local variable to `sro` which changes tons of lines. First, what's the reasoning behind that change in naming -- is that an upstream change, and is `sro` the right name, or something like `s3_resource`? And in any case, I think we'd be a lot better off doing that rename in a separate commit either before or after the big version upgrade.

Yep that's an upstream change, I'll extract that out to another commit after the upgrade.

whoodes commented:
K I combed through this and made it a lot easier to review hopefully.  There were a lot of little changes to the original logic as you said, and said changes have all been reverted.  This commit should now purely focus on the upgrade.

One thing I'm still unsure about is the `self.connection.DefaultHost` change in `upload.py`.  Having `region_name` actually broke things when I put them back how they were supposed to be (eg  fixing the `get_bucket` method).  I've scoured the docs, but have failed to find any satisfying results.

timabbott commented on zerver/tests/test_transfer.py:
```diff
@@ -41,16 +41,16 @@ def test_transfer_avatars_to_s3(self) -> None:
-        self.assertEqual(image_key.get_contents_as_string(), open(avatar_disk_path(user), "rb").read())
-        self.assertEqual(original_image_key.get_contents_as_string(), open(avatar_disk_path(user, original=True), "rb").read())
-        self.assertEqual(medium_image_key.get_contents_as_string(), open(avatar_disk_path(user, medium=True), "rb").read())
+        self.assertEqual(sum(1 for _ in bucket.objects.all()), 3)
```
`len(list(bucket.objects.all()))` is the right way to do this, rather than this weird `sum(1...` game.

andersk commented on requirements/prod.txt:
```diff
@@ -43,9 +43,12 @@ beautifulsoup4==4.8.2 \
-boto==2.49.0 \
-    --hash=sha256:147758d41ae7240dc989f0039f27da8ca0d53734be0eb869ef16e3adcfa462e8 \
-    --hash=sha256:ea0d3b40a2d852767be77ca343b58a9e3a4b00d9db440efb8da74b4e58025e5a
+boto3==1.12.31 \
+    --hash=sha256:913fc7bbb9df147ed6fa0bd2b391469652ee8cad3e26ca2355e6ff774d9516fb
+botocore==1.15.31 \
+    --hash=sha256:3d4684f61ff07aa1b4cd30d13a6b8e6416142be5a2a55aedcc1d7974b9415bb1 \
+    --hash=sha256:86e3f25c9a94e9e85915ef3dcb13b496d25873a40b81f4e23a7a0250248d1be9 \
+    # via boto3, s3transfer
```
You ended up with a different version here than in `dev.txt`, which could invite tricky production-only bugs. (We should maybe come up with a better way to avoid and test for this kind of skew.)

timabbott commented:
OK I made the following changes while editing this:
* Re-added the progress callback that was removed when uploading the giant export tarball.
* Fixed the weird `sum(1 ...` pattern for getting length of an iterable.
* Tweaked some comments on tests.

And am pushing back to this PR.  Changes below:

```
(zulip-py3-venv) tabbott@coset:~/zulip$ git diff | cat
diff --git a/zerver/lib/upload.py b/zerver/lib/upload.py
index 14d9bc5a17..6eb5aeb47a 100644
--- a/zerver/lib/upload.py
+++ b/zerver/lib/upload.py
@@ -1,4 +1,4 @@
-from typing import Optional, Tuple
+from typing import Any, Optional, Tuple
 
 from django.utils.translation import ugettext as _
 from django.conf import settings
@@ -35,6 +35,7 @@ import io
 import random
 import logging
 import shutil
+import sys
 
 DEFAULT_AVATAR_SIZE = 100
 MEDIUM_AVATAR_SIZE = 500
@@ -611,12 +612,16 @@ class S3UploadBackend(ZulipUploadBackend):
         return "https://%s.s3.amazonaws.com/%s" % (bucket, emoji_path)
 
     def upload_export_tarball(self, realm: Optional[Realm], tarball_path: str) -> str:
+        def percent_callback(bytes_transferred: Any) -> None:
+            sys.stdout.write('.')
+            sys.stdout.flush()
+
         session = boto3.Session(settings.S3_KEY, settings.S3_SECRET_KEY)
         # We use the avatar bucket, because it's world-readable.
         bucket = get_bucket(session, settings.S3_AVATAR_BUCKET)
         key = bucket.Object(os.path.join("exports", generate_random_token(32),
                             os.path.basename(tarball_path)))
-        key.upload_file(tarball_path)
+        key.upload_file(tarball_path, Callback=percent_callback)
 
         session = botocore.session.get_session()
         config = Config(signature_version=botocore.UNSIGNED)
diff --git a/zerver/tests/test_transfer.py b/zerver/tests/test_transfer.py
index a4537336a0..41b12318be 100644
--- a/zerver/tests/test_transfer.py
+++ b/zerver/tests/test_transfer.py
@@ -45,7 +45,7 @@ class TransferUploadsToS3Test(ZulipTestCase):
         original_image_key = bucket.Object(path_id + ".original")
         medium_image_key = bucket.Object(path_id + "-medium.png")
 
-        self.assertEqual(sum(1 for _ in bucket.objects.all()), 3)
+        self.assertEqual(len(list(bucket.objects.all())), 3)
         self.assertEqual(image_key.get()['Body'].read(), open(avatar_disk_path(user), "rb").read())
         self.assertEqual(original_image_key.get()['Body'].read(), open(avatar_disk_path(user, original=True), "rb").read())
         self.assertEqual(medium_image_key.get()['Body'].read(), open(avatar_disk_path(user, medium=True), "rb").read())
@@ -63,7 +63,7 @@ class TransferUploadsToS3Test(ZulipTestCase):
 
         attachments = Attachment.objects.all()
 
-        self.assertEqual(sum(1 for _ in bucket.objects.all()), 2)
+        self.assertEqual(len(list(bucket.objects.all())), 2)
         self.assertEqual(bucket.Object(attachments[0].path_id).get()['Body'].read(), b'zulip1!')
         self.assertEqual(bucket.Object(attachments[1].path_id).get()['Body'].read(), b'zulip2!')
 
@@ -87,7 +87,7 @@ class TransferUploadsToS3Test(ZulipTestCase):
 
         transfer_emoji_to_s3(1)
 
-        self.assertEqual(sum(1 for _ in bucket.objects.all()), 2)
+        self.assertEqual(len(list(bucket.objects.all())), 2)
         original_key = bucket.Object(emoji_path + ".original")
         resized_key = bucket.Object(emoji_path)
 
diff --git a/zerver/tests/test_upload.py b/zerver/tests/test_upload.py
index 3bdf1efb56..cfffc79a34 100644
--- a/zerver/tests/test_upload.py
+++ b/zerver/tests/test_upload.py
@@ -1577,6 +1577,9 @@ class S3Test(ZulipTestCase):
         response = self.client_get(uri)
         self.assertEqual(response.status_code, 302)
 
+        # TODO boto3 upgrade: This is a broken way to check if
+        # the key from our redirect is in boto; we are no longer
+        # verifying the redirect is to the target file :( :( :(
         for key in bucket.objects.all():
             key = key.key
         self.assertIsNotNone(key)
@@ -1611,7 +1614,10 @@ class S3Test(ZulipTestCase):
         self.assertEqual(medium_image_key.key, medium_path_id)
         medium_image_data = medium_image_key.get()['Body'].read()
         self.assertEqual(medium_image_data, test_medium_image_data)
-        # Boto3 requires us to iterate through the keys
+
+        # TODO boto3: The next is intended to just delete medium_image_key
+        # This is a hacky Boto3 way to do so by iterating through the keys;
+        # there is surely a better way to do this.
         for key in bucket.objects.all():
             if key is medium_image_key:
                 key.delete()  # nocoverage
@@ -1685,8 +1691,8 @@ class S3Test(ZulipTestCase):
         zerver.lib.actions.do_delete_avatar_image(user)
 
         self.assertEqual(user.avatar_source, UserProfile.AVATAR_FROM_GRAVATAR)
-        # Boto3 won't return a None type from trying to get a non-existent key,
-        # it just raises a ClientError
+
+        # Confirm that the avatar files no longer exist in S3.
         with self.assertRaises(botocore.exceptions.ClientError):
             bucket.Object(avatar_path_id).load()
         with self.assertRaises(botocore.exceptions.ClientError):
```

timabbott commented:
I opened https://github.com/zulip/zulip/issues/14484 to cover most of the follow-up regressions/investigations we'll want to do after merging this; I think all of those are things that can wait a week or two after this merges.  

Before merging this, I'd like us to:
* Fix the version skew issue Anders pointed out
* Ideally, fix the two tests that I added TODO comments on in the last remarks.
* @whoodes have you tested locally in a development environment with the S3 backend setup that all the main code paths work?  It's not hard to do if you just create a free S3 account and then configure the settings in `dev_settings.py` etc. locally.).
* Test that everything works correctly manually on an S3 production system (I can probably take care of this on one of our staging servers;

whoodes commented:
> @whoodes have you tested locally in a development environment with the S3 backend setup that all the main code paths work? It's not hard to do if you just create a free S3 account and then configure the settings in dev_settings.py etc. locally.).

I have not.  I didn't know it was that easy, so yeah, no problem.

> verifying the redirect is to the target file :( :( :(

Oh man! Triple sad face, that is not good.  I'll definitely work fix up the tests.

> Fixed the weird sum(1 ... pattern for getting length of an iterable.

Haha, sometimes I need to remember to use common sense instead of google.

> @whoodes can you do any updates to this PR as new fixup commits? We'll squash before we merge, but it's most convenient if I can easily see changes from what I just reviewed.

No problem, @timabbott  do you want the version skew fix in a new commit as well, or should I just amend this one for that issue in particular?

timabbott commented:
> No problem, @timabbott do you want the version skew fix in a new commit as well, or should I just amend this one for that issue in particular?

Doesn't super matter; that piece is fine to squash.

timabbott commented:
Let me know when you have this ready!

whoodes commented:
Updated :) 
* Fixed those two TODOs.
* I'm not sure what to do about the boto3 version.  If I don't pin down a version number, I get the skew between dev and prod.
* Tested the S3 stuff as mentioned in chat.

andersk commented:
> I'm not sure what to do about the boto3 version. If I don't pin down a version number, I get the skew between dev and prod.

I’ve now eliminated the possibility of that skew by always generating the prod lock as a subset of the dev lock (21c2a4cc42447f5380c2b2d5638d7505658a52c0, #14621).

timabbott commented:
Nice.  @whoodes can you rebase?

whoodes commented:
Awesome!  I've rebased.

whoodes commented:
K, this has been rebased.

timabbott commented:
I merged the migration deprecation commits after a commit message tweak; with us having dropped Xenial support, it's extra clear they're safe.

timabbott commented:
I squashed the squash commits and pushed back here.

One question: Why does this use `verify=True` in some places, and False in others?

whoodes commented:
I read in the AWS docs for a replacement to `validate` in boto3, and looking back at things again, I may have misread something.  [`verify`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html#boto3.session.Session.client) is for verifying ssl certificates, which I presumably associated with https://aws.amazon.com/blogs/developer/configure-boto-to-validate-https-certificates/  . But we are only validating buckets here, which doesn't seem to be an issue anymore, as mentioned here, https://github.com/zulip/zulip/pull/10986#issuecomment-537226703.

So, we can probably just remove those actually.

andersk commented on zerver/lib/upload.py:
```diff
@@ -269,16 +272,10 @@ def realm_avatar_and_logo_path(self, realm: Realm) -> str:
+def get_bucket(session: Session, bucket_name: str) -> ServiceResource:
+    # See https://github.com/python/typeshed/issues/2706
+    # for why this return type is a `ServiceResource`.
+    bucket = session.resource('s3', verify=False).Bucket(bucket_name)
```
Yeah, we should remove `verify` options so we don’t disable SSL verification. Line commenting to make sure we don’t forget.

The unrelated Boto 2 `validate=False` option appears to be the default behavior of Boto 3.

whoodes replied:
Cool, thanks Anders. And my apology for the oversight.

timabbott commented:
@whoodes let me know when you've addressed that `verify` comment and rebased this.

whoodes commented:
K, rebased and got rid of the incorrect SSL verifications.

timabbott commented:
Tested this in production and confirmed it's working, so merged, thanks @whoodes!

There's one follow-up task for this, which is that we need to determine the right way to use boto3 to get the URL in the presence of third-party S3-compatible providers, to replace this code:

```
-        return "https://%s.%s/%s/realm/%s?version=%s" % (
-            bucket, self.connection.DefaultHost, realm_id, file_name, version)
+        return "https://%s.s3.amazonaws.com/%s/realm/%s?version=%s" % (
+            bucket, realm_id, file_name, version)
```

@whoodes can you open an issue for that?

timabbott commented:
Looks like `bucket.meta.client.meta.endpoint_url` does this:

```
In [9]: bucket.meta.client.meta.endpoint_url                                                
Out[9]: 'https://s3.amazonaws.com'
```

So I think we can mostly do that, though we may need to refactor things a bit to only initialize each Bucket object once (rather than once per call to `get_avatar_url`, for example, which would be very expensive)

whoodes commented:
Great! I'll open up an issue to track that.

whoodes commented:
Opened #15373 for the above.
